x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,216 @@
+
+		for i := range uniqueSlots {
+			if slots[k].Equal(uniqueSlots[i]) {
+				uniqueSlots[i].Count += 1
+				isUnique = false
+				break
+			}
+		}
+
+		if isUnique {
+			uniqueSlots = append(uniqueSlots, slots[k])
+		}
+	}
+
+	return uniqueSlots
+}
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func GetStrategy() ScoreStrategy {
+	return uniformStrategy
+}
+
+// CalcPairingScore calculates the final pairing score for a pairing slot (with strategy)
+// For efficiency purposes, we calculate the score on a diff slot which represents the diff reqs of the current slot
+// and the previous slot
+func CalcPairingScore(scores []*PairingScore, strategy ScoreStrategy, diffSlot *PairingSlot, minStake sdk.Int) error {
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight := strategy[reqName] // not checking if key is found because it's verified in init()
+
+		for _, score := range scores {
+			newScoreComp := req.Score(*score.Provider)
+			if newScoreComp == 0 {
+				err := fmt.Errorf("score component is zero. score component name: %s, provider address: %s", reqName, score.Provider.Address)
+				panic(err)
+			}
+			newScoreComp = commontypes.SafePow(newScoreComp, weight)
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				score.Score /= prevReqScoreComp
+			}
+			score.Score *= newScoreComp
+
+			// update the score component map
+			score.ScoreComponents[reqName] = newScoreComp
+		}
+	}
+
+	return nil
+}
+
+// PrepareHashData prepares the hash needed in the pseudo-random choice of providers
+func PrepareHashData(projectIndex string, chainID string, epochHash []byte) []byte {
+	hashData := []byte{}
+
+	// add the session start block hash to the function to make it as unpredictable as we can
+	hashData = append(hashData, epochHash...)
+	hashData = append(hashData, chainID...)      // to make this pairing unique per chainID
+	hashData = append(hashData, projectIndex...) // to make this pairing unique per consumer
+
+	return hashData
+}
+
+// PickProviders pick a <group-count> providers set with a pseudo-random weighted choice (using the providers' score list and hashData)
+func PickProviders(ctx sdk.Context, scores []*PairingScore, groupCount int, hashData []byte, chosenProvidersIdx map[int]bool) (returnedProviders []epochstoragetypes.StakeEntry) {
+	if len(scores) == 0 {
+		return returnedProviders
+	}
+
+	var scoreSum uint64
+	for idx, providerScore := range scores {
+		if chosenProvidersIdx[idx] {
+			// skip index of providers already selected
+			continue
+		}
+		scoreSum += providerScore.Score
+	}
+	if scoreSum == 0 {
+		err := fmt.Errorf("score sum is zero. Cannot pick providers for pairing")
+		panic(err)
+	}
+
+	iterationNum := commontypes.FindMin([]int{groupCount, len(scores) - len(chosenProvidersIdx)})
```

Comment:
instead of this you can do:
if groupCount >= len(scores) - len(chosenProvidersIdx) //this means we will want every provider so no need to calculate
    return all providers

Reply:
I can do this, but I'll need to write another loop (because I've got the scores and not providers. So I'll need to create a new array and append providers from scores). What's wrong with the current code?

Reply:
you will do computations of scoring and choosing for nothing instead of returning all of them

Reply:
done. see https://github.com/lavanet/lava/commit/6ba55d5846fd3f875247146f357d52c5249e7d4a